src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs

```diff
@@ -747,8 +747,22 @@ public BoundExpression Sequence(params BoundExpression[] parts)
         public BoundExpression Sequence(ImmutableArray<LocalSymbol> locals, params BoundExpression[] parts)
         {
             var builder = ArrayBuilder<BoundExpression>.GetInstance();
-            for (int i = 0; i < parts.Length - 1; i++) builder.Add(parts[i]);
+            for (int i = 0; i < parts.Length - 1; i++)
+            {
+                var part = parts[i];
+                if (LocalRewriter.ReadIsSideeffecting(part))
```

Comment:
>if (LocalRewriter.ReadIsSideeffecting(part)) [](start = 16, length = 44)

This doesn't feel like the right place to do optimizations like this. Can't we do this in optimizer/emitter?